components/brave_rewards/renderer/creator_detection_agent.cc

```diff
@@ -0,0 +1,265 @@
+  if (!origin.is_valid() || !origin.SchemeIs(url::kHttpsScheme) ||
+      !IsMediaPlatformURL(origin)) {
+    return std::nullopt;
+  }
+
+  auto iter = kScriptMap.find(origin.host_piece());
+  if (iter == kScriptMap.end()) {
+    return std::nullopt;
+  }
+
+  return DetectionScript{
+      .source = blink::Platform::Current()->GetDataResourceString(iter->second),
+      .should_run_in_isolated_world = ShouldRunInIsolatedWorld(iter->second)};
+}
+
+v8::MaybeLocal<v8::Function> CompileDetectionInitializer(
+    v8::Local<v8::Context> context,
+    std::string_view function_body) {
+  v8::Context::Scope context_scope(context);
+  v8::MicrotasksScope microtasks_scope(
+      context, v8::MicrotasksScope::kDoNotRunMicrotasks);
+
+  auto* isolate = context->GetIsolate();
+  v8::ScriptCompiler::Source source(gin::StringToV8(isolate, function_body));
+
+  v8::Local<v8::String> parameters[] = {
+      gin::StringToV8(isolate, "setPageChangedCallback"),
+      gin::StringToV8(isolate, "onCreatorDetected"),
+      gin::StringToV8(isolate, "verboseLogging")};
+
+  // The creator detection script is compiled as a function body. Browser
+  // capabilities are passed in to the compiled function as arguments whose
+  // parameter names are defined above. Browser capabilities are only exposed
+  // to the script; they are not visible outside of the compiled function.
+  return v8::ScriptCompiler::CompileFunction(context, &source, 3, parameters);
+}
+
+template <typename... Args>
+void CallFunctionWithArgs(v8::Isolate* isolate,
+                          v8::Local<v8::Function> function,
+                          Args... args) {
+  v8::Local<v8::Context> context = function->GetCreationContextChecked(isolate);
+  v8::Context::Scope context_scope(context);
+  v8::MicrotasksScope microtasks_scope(
+      context, v8::MicrotasksScope::kDoNotRunMicrotasks);
+  std::array<v8::Local<v8::Value>, sizeof...(Args)> arg_list = {
+      args.template As<v8::Value>()...};
+  std::ignore = function->Call(context, context->Global(), arg_list.size(),
+                               arg_list.empty() ? nullptr : arg_list.data());
+}
+
+}  // namespace
+
+CreatorDetectionAgent::CreatorDetectionAgent(content::RenderFrame* render_frame,
+                                             int32_t isolated_world_id)
+    : RenderFrameObserver(render_frame),
+      isolated_world_id_(isolated_world_id) {}
+
+CreatorDetectionAgent::~CreatorDetectionAgent() = default;
+
+void CreatorDetectionAgent::DidCommitProvisionalLoad(
+    ui::PageTransition transition) {
+  if (render_frame()->IsMainFrame()) {
+    MaybeInjectDetectionScript();
+  }
+}
+
+void CreatorDetectionAgent::DidFinishSameDocumentNavigation() {
+  if (render_frame()->IsMainFrame()) {
+    NotifyPageChanged();
+  }
+}
+
+void CreatorDetectionAgent::WillReleaseScriptContext(
+    v8::Local<v8::Context> context,
+    int32_t world_id) {
+  if (page_changed_callback_.IsEmpty()) {
```

Comment:
I wanted to ensure that it was not possible to accidentally call `page_changed_callback_` after the context was released. It's not completely obvious to me that that is not possible if we don't reset it here. But maybe there's some ordering invariants that I'm not aware of.

Reply:
I believe `DidClearWindowObject` is normally used for cleanup/setup
ex:

```
void JsCommunication::DidClearWindowObject() {
  if (inside_did_clear_window_object_)
    return;

  base::AutoReset<bool> flag_entry(&inside_did_clear_window_object_, true);

  // Invalidate `weak_ptr_factory_for_bindings_` so that existing bindings
  // can not send messages back to the browser (JsBinding is owned by v8,
  // so we can't delete it here).
  weak_ptr_factory_for_bindings_.InvalidateWeakPtrs();
```